### PR TITLE
feat[venom]: improve unused variable removal pass

### PR DESCRIPTION
**Pull request by harkal**
### What I did

Implemented a better variable elimination pass. This new implementation is efficient and manages to remove all unused variables in one run.

Additionally, this PR adds some extra utility functions to `OrderedSet()`

### Commit message

```
This commit improves the variable elimination pass by efficiently removing 
all instructions that produce unused output. This new algorithm removes all
unused variables in one run.

Additionally, it adds the utility functions `pop()` and `addmany()` to the
`OrderedSet()` class.

The `DFTAnalysis` is also augmented with a method to remove uses:
`remove_use(self, op: IRVariable, inst: IRInstruction)`
```

### Cute Animal Picture

```
                 __
                / _)--< 
       _.----._/ /          
     /          /  Dinosnek          
 __/ (  |  (   |             
/__.-'|_|--|__|             
```

### CODE REVIEW

File: vyper/utils.py [hunk 1]
```diff
@@ -45,9 +45,16 @@ def __len__(self):
     def first(self):
         return next(iter(self))
 
+    def pop(self):
```
Comment (charles-cooper) [inline vyper/utils.py:48]: Maybe popleft()? Since it's lifo

File: vyper/venom/passes/remove_unused_variables.py [hunk 2]
```diff
@@ -1,47 +1,42 @@
...
+        if len(uses) > 0:
+            return
 
-            # backtrack to the *previous* instruction, in case we removed
-            # an instruction which had prevented the previous instruction
-            # from being removed
-            i -= 2
+        for operand in inst.get_inputs():
+            new_uses = self.dfg.remove_use(operand, inst)
+            for use in new_uses:
```
Comment (charles-cooper) [inline vyper/venom/passes/remove_unused_variables.py:39]: can use `self.work_list.addmany(new_uses)`

File: vyper/venom/analysis/dfg.py [hunk 3]
```diff
@@ -22,6 +22,11 @@ def get_uses(self, op: IRVariable) -> list[IRInstruction]:
     def get_producing_instruction(self, op: IRVariable) -> Optional[IRInstruction]:
         return self._dfg_outputs.get(op)
 
+    def remove_use(self, op: IRVariable, inst: IRInstruction) -> list[IRInstruction]:
+        uses = self._dfg_inputs.get(op, [])
+        uses.remove(inst)
+        return uses
```
Comment (charles-cooper) [inline vyper/venom/analysis/dfg.py:28]: nit: i'd prefer an API which does one thing at a time, so `remove_use()` just removes the item, and then the caller needs another call to `.get_uses()`

**Review by charles-cooper — COMMENTED** [review]
nice algorithm :) looks like it's faster too! left a couple nits, otherwise looks good to merge

File: vyper/venom/passes/remove_unused_variables.py (hunk 2 above)
Comment (harkal) [reply vyper/venom/passes/remove_unused_variables.py:39]: Even I can’t keep up with all the cool utility functions I write 😅

File: vyper/venom/analysis/dfg.py (hunk 3 above)
Comment (harkal) [reply vyper/venom/analysis/dfg.py:28]: It’s a pattern I use regularly, but I have no strong preference about doing it in two steps

File: vyper/utils.py (hunk 1 above)
Comment (harkal) [reply vyper/utils.py:48]: popitem() is also lifo

Comment (charles-cooper) [reply vyper/utils.py:48]: > popitem() is also lifo

right, `pop()` (on lists) is FIFO though

File: vyper/venom/passes/remove_unused_variables.py [hunk 4]
```diff
@@ -1,47 +1,42 @@
...
-            if inst.volatile:
-                continue
+        work_list = OrderedSet()
+        self.work_list = work_list
 
-            # Skip instructions without output
-            if inst.output is None:
-                continue
+        for _, inst in self.dfg.outputs.items():
+            work_list.add(inst)
```
Comment (charles-cooper) [inline vyper/venom/passes/remove_unused_variables.py:22]:
```suggestion
        uses = self.dfg.outputs.values()
        work_list.addmany(uses)
```

Comment (charles-cooper) [reply vyper/venom/passes/remove_unused_variables.py:22]: i'm ok with either way tbh

File: vyper/utils.py (hunk 1 above)
Comment (charles-cooper) [reply vyper/utils.py:48]:
```suggestion
    def popleft(self):
```

Comment (harkal) [reply vyper/utils.py:48]: Seems to me that `pop()` is also LIFO on lists... :confusion:

Comment (charles-cooper) [reply vyper/utils.py:48]: idts, there is `pop()` for lists which is FIFO, and `popleft()` for deques which is LIFO

Comment (charles-cooper) [reply vyper/utils.py:48]: whoops, i switched FIFO and LIFO

**Review by charles-cooper — APPROVED** [review]

**Pull request merged**